addons/web/static/src/js/core/popover.js

```diff
@@ -26,37 +26,75 @@ odoo.define('web.Popover', function () {
             this.state = useState({
                 displayed: false,
             });
+
+            this._onDocumentClick = this._onDocumentClick.bind(this);
+            this._onDocumentScroll = this._onDocumentScroll.bind(this);
+            this._onWindowResize = this._onWindowResize.bind(this);
+
+            this._onDocumentScroll = _.throttle(this._onDocumentScroll, 50);
+            this._onWindowResize = _.debounce(this._onWindowResize, 250);
+        }
+
+        mounted() {
+            /**
+             * Use capture for the following events to ensure no other part of
+             * the code can stop its propagation from reaching here.
+             */
+            document.addEventListener('click', this._onDocumentClick, {
+                capture: true,
+            });
+            document.addEventListener('scroll', this._onDocumentScroll, {
+                capture: true,
+            });
+            window.addEventListener('resize', this._onWindowResize);
+
+            this._compute();
         }
 
         patched() {
-            super.patched(...arguments);
-            this.show();
+            this._compute();
+        }
+
+        willUnmount() {
+            document.removeEventListener('click', this._onDocumentClick, {
+                capture: true,
+            });
+            document.removeEventListener('scroll', this._onDocumentScroll, {
+                capture: true,
+            });
+            window.removeEventListener('resize', this._onWindowResize);
         }
 
+        //----------------------------------------------------------------------
+        // Private
+        //----------------------------------------------------------------------
+
         /**
-         * Show the popover according to its props. This method will try to position the
+         * Computes the popover according to its props. This method will try to position the
          * popover as requested (according to the `position` props). If the requested position
          * does not fit the viewport, other positions will be tried in a clockwise order starting
          * a the requested position (e.g. starting from left: top, right, bottom). If no position
          * is found that fits the viewport, 'bottom' is used.
          */
-        show() {
+        _compute() {
             if (!this.state.displayed) {
                 return;
             }
+
             // get the target from the dom
             // we don't want to do this early, since the target might not
-            // be in the dom when the component is instanciated
-            if (!this.constructor._isInViewport(this.el)) {
+            // be in the dom when the component is instantiated
+            if (!this.env.services.isElementInViewport(this.el)) {
```

Comment:
weird to define that as a service, but I suppose this is to easily mock it in tests and remove the hard-code styling of before.

Reply:
This is how the `web` env is currently set up, "helpers" are basically all in `services`.

As for the goal:
- it shouldn't be in the component itself as it is a fairly generic method
- it could be a helper, but it is harder to test
- -> env is the best place for it

One of the tests is still relying on the hard-coded positioning (the one actually testing the position of the popover in regard to the viewport) but all the other tests indeed don't care about this at all, as long as a popover is shown.